File: backend/src/test/java/com/gridsync/crdt/CrdtMergerTest.java
package com.gridsync.crdt;

import org.junit.jupiter.api.Test;
import java.util.UUID;
import java.lang.reflect.Method;
import static org.junit.jupiter.api.Assertions.*;

class CrdtMergerTest {

    @Test
    void testMergeNormalOrdering() {
        UUID replicaA = UUID.randomUUID();
        UUID replicaB = UUID.randomUUID();
        
        CellValue valA = new CellValue("A", new HybridLogicalClock(1000, 0, replicaA), replicaA);
        CellValue valB = new CellValue("B", new HybridLogicalClock(1001, 0, replicaB), replicaB);
        
        // valB has higher physical time, so it should win
        assertEquals(valB, CrdtMerger.merge(valA, valB));
        assertEquals(valB, CrdtMerger.merge(valB, valA));
        
        CellValue valC = new CellValue("C", new HybridLogicalClock(1001, 1, replicaA), replicaA);
        
        // valC has same physical time but higher logical counter, should win
        assertEquals(valC, CrdtMerger.merge(valB, valC));
    }

    @Test
    void testMergeTieOnIdenticalHLC() {
        UUID replicaA = new UUID(0, 1);
        UUID replicaB = new UUID(0, 2); // B is strictly greater than A
        
        // Even though HLC replicaIds are the same (which might happen if HLC didn't include it or it's forced),
        // we'll make the HLCs exactly identical.
        HybridLogicalClock hlcA = new HybridLogicalClock(1000, 0, replicaA);
        HybridLogicalClock hlcB = new HybridLogicalClock(1000, 0, replicaA); // Identical HLC
        
        CellValue valA = new CellValue("A", hlcA, replicaA);
        CellValue valB = new CellValue("B", hlcB, replicaB);
        
        // HLC is identical, so it falls back to CellValue.replicaId. replicaB > replicaA, so valB wins.
        assertEquals(valB, CrdtMerger.merge(valA, valB));
        assertEquals(valB, CrdtMerger.merge(valB, valA));
    }
    
    @Test
    void testMergeTieOnIdenticalHLCAndIdenticalReplicaId() {
        UUID replicaA = UUID.randomUUID();
        
        HybridLogicalClock hlcA = new HybridLogicalClock(1000, 0, replicaA);
        
        CellValue valA = new CellValue("A", hlcA, replicaA);
        CellValue valB = new CellValue("B", hlcA, replicaA);
        
        // Should not crash, should just return one of them (valB in this case based on cmp > 0 ? a : b)
        // Since cmp is 0 and replica.compareTo is 0, it evaluates to false, returning b (valB).
        assertDoesNotThrow(() -> {
            CellValue result = CrdtMerger.merge(valA, valB);
            assertNotNull(result);
        });
    }

    @Test
    void testMergeHasNoServerTimeDependency() throws NoSuchMethodException {
        // Asserting that the signature is strictly public static CellValue merge(CellValue a, CellValue b)
        Method mergeMethod = CrdtMerger.class.getMethod("merge", CellValue.class, CellValue.class);
        assertNotNull(mergeMethod);
        
        // Assert no other parameters are accepted
        assertEquals(2, mergeMethod.getParameterCount());
        
        // Ensure no overloaded methods exist that might take time/clock
        for (Method m : CrdtMerger.class.getDeclaredMethods()) {
            if (m.getName().equals("merge")) {
                assertEquals(2, m.getParameterCount(), "merge() should only accept 2 CellValues");
            }
        }
    }
}

File: frontend/src/crdt/__tests__/CrdtMerger.test.ts
import { describe, it, expect } from 'vitest';
import { merge } from '../CrdtMerger';
import { CellValue } from '../CellValue';
import { HybridLogicalClock } from '../HybridLogicalClock';
import { v4 as uuidv4 } from 'uuid';

describe('CrdtMerger', () => {
  it('merge normal ordering', () => {
    const replicaA = uuidv4();
    const replicaB = uuidv4();

    const valA: CellValue = { value: 'A', hlc: new HybridLogicalClock(1000, 0, replicaA), replicaId: replicaA };
    const valB: CellValue = { value: 'B', hlc: new HybridLogicalClock(1001, 0, replicaB), replicaId: replicaB };

    expect(merge(valA, valB)).toBe(valB);
    expect(merge(valB, valA)).toBe(valB);

    const valC: CellValue = { value: 'C', hlc: new HybridLogicalClock(1001, 1, replicaA), replicaId: replicaA };
    expect(merge(valB, valC)).toBe(valC);
  });

  it('merge tie on identical HLC', () => {
    const replicaA = "11111111-1111-1111-1111-111111111111";
    const replicaB = "22222222-2222-2222-2222-222222222222";

    const hlcA = new HybridLogicalClock(1000, 0, replicaA);
    const hlcB = new HybridLogicalClock(1000, 0, replicaA); // Identical HLC intentionally

    const valA: CellValue = { value: 'A', hlc: hlcA, replicaId: replicaA };
    const valB: CellValue = { value: 'B', hlc: hlcB, replicaId: replicaB };

    // B's replicaId is larger than A's replicaId lexicographically, so B wins tiebreak
    expect(merge(valA, valB)).toBe(valB);
    expect(merge(valB, valA)).toBe(valB);
  });

  it('merge tie on identical HLC and identical replicaId', () => {
    const replicaA = uuidv4();
    const hlcA = new HybridLogicalClock(1000, 0, replicaA);

    const valA: CellValue = { value: 'A', hlc: hlcA, replicaId: replicaA };
    const valB: CellValue = { value: 'B', hlc: hlcA, replicaId: replicaA };

    // Should not crash, just returns one of them deterministically
    const result = merge(valA, valB);
    expect(result).toBeDefined();
  });
});
